Add a key binding Toggle Acrylic

Jaswir opened the pull request:
## Summary of the Pull Request
Added a key binding Toggle Acrylic 

## References and Relevant Issues
#2531

## Detailed Description of the Pull Request / Additional comments
![toggle acrylic v1](https://github.com/microsoft/terminal/assets/15957528/65d626d8-c4e8-494d-93d3-59d8099492e4)

## Validation Steps Performed
Tested whether things are still working along with:
* Mouse scroll wheel
* Adjust opacity: ![image](https://github.com/microsoft/terminal/assets/15957528/0b834eaa-d0f8-4448-8a95-560d41848556)
* Acrylic Off when they turn off transparency.

In both situations with 
![image](https://github.com/microsoft/terminal/assets/15957528/1859287c-389b-4ead-9d20-81f6baff5ef9)

```
"useAcrylic": true 
"useAcrylic": false
```

## PR Checklist
- [ V] Closes #2531 
- [X ] Tests added/passed
- [ X] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
- [X] Schema updated (if necessary)

Jaswir committed:
Rebase 2531 Add a key binding to set or toggle acrylic to Master

Jaswir commented:
@zadjii-msft 

Why do I get the formatting error? I thought it was that I needed to remove the `\`
`    ON_ALL_ACTIONS(RestartConnection)       \`

zadjii-msft commented on src/cascadia/TerminalSettingsModel/AllShortcutActions.h:
```diff
@@ -106,7 +107,7 @@
     ON_ALL_ACTIONS(ExpandSelectionToWord)   \
     ON_ALL_ACTIONS(CloseOtherPanes)         \
-    ON_ALL_ACTIONS(RestartConnection)
+    ON_ALL_ACTIONS(RestartConnection)       
```
ah, I bet it's complaining about the trailing spaces here. 


```
tools\runformat.cmd
```

should automatically reformat everything for you 😉

Jaswir replied:
![image](https://github.com/microsoft/terminal/assets/15957528/5a8cecb0-bb48-49fb-96d4-be50afe69e4b)

Jaswir replied:
> ah, I bet it's complaining about the trailing spaces here.
> 
> ```
> tools\runformat.cmd
> ```
> 
> should automatically reformat everything for you 😉

Things are never easy are they. 
How can I solve. Noob at ps1 scripts but It seems like it should work cause the Invoke-CodeFormat function is inside the OpenConsole.psm1 script. As for now I went the manual route. Problem seems to be in ControlCore.cpp

DHowett replied:
Sorry about that.

Can you do this instead? It lets you do the steps from "runformat" manually.

It will be much easier than trying all your individual files to see which one was the problem.

```
import-module ./tools/openconsole.psm1
invoke-codeformat -ignorexaml
```

it will make changes to your code which you will need to commit. :smile:

Jaswir replied:
@zadjii-msft @DHowett 

I solved the issues. It's ready to be reviewed now.

Jaswir committed:
Run Code Formatting

Jaswir commented:
Solved the issues. It's ready to be reviewed. Let me know if any changes need to be made.

Jaswir commented:
@zadjii-msft @DHowett-MSFT @DHowett 

Hi Dustin, Hi Mike, 

I was wondering if you got a chance yet to review the code? 

@zadjii-msft 
Or would it stop being able to talk with you in a more comfortable manner about the other issue inside 2531, because you wouldn't get notifications anymore once it's merged or something?

zadjii-msft commented on src/cascadia/TerminalControl/ControlCore.h:
```diff
@@ -272,6 +274,8 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             std::shared_ptr<ThrottledFuncTrailing<Control::ScrollPositionChangedArgs>> updateScrollBar;
         };
 
+        bool _acrylicToggle;
```
Why add another bool here, rather than leverage the existing `UseAcrylic` `RUNTIME_SETTING`?

Jaswir replied:
Initially I used the `UseAcrylic RUNTIME_SETTING `and it seemed to work, but after testing Mouse Scroll Wheel to change Opacity and Adjust Opacity to change opacity, I stumbled upon some problems with this method.

**Problems:**
![runtime_acrylic_breaks_adjustopac](https://github.com/microsoft/terminal/assets/15957528/fb36cf23-3c7b-4cb8-bf55-7879260aac51)
![runtime_acrylic_breaks_scrollwheel](https://github.com/microsoft/terminal/assets/15957528/f000a464-932a-4a80-966a-bbd78424c1fc)

**The actual Terminal Preview behaviour with UseAcrylic on in settings.json in these scenarios:**
![runtime_acrylic_correct_adjustopacity](https://github.com/microsoft/terminal/assets/15957528/a1912424-9661-4a57-8f8d-9d93dd5cd4e6)
![runtime_acrylic_correct_scrollwheel](https://github.com/microsoft/terminal/assets/15957528/6da920b5-6fba-43e9-b3aa-e3d60c3e2d16)

When the user uses Mouse Scroll Wheel to change Opacity and Adjust Opacity to change opacity through the command palette the `ControlCore::_setOpacity `function is triggered. This function manually turns off the` _runtimeUseAcrylic` when opacity becomes 100. 

When the user scroll to opacity 100 and then scroll down again, the runtime acrylic goes false at 100 and would stay false when you go back under 100. Also when adjusting the opacity through the command palette similar thing happens when going to set background opacity to 100% option it stays false when going to 25%, 50%, etc.

Basically we can’t use the` UseAcrylic RUNTIME_SETTING`, because it becomes False when the user turns off transparency(opacity becomes 100)  by using Mouse Scroll Wheel to change Opacity and Adjust Opacity to change opacity and we can’t use `UseAcrylic RUNTIME_SETTING` to turn `_runtimeUseAcrylic `back on since it’s false. So I figured we need to use an external boolean besides the `UseAcrylic () RUNTIME_SETTING`.

Using an external boolean solves these issues.

Jaswir replied:
How would you have solved this problem?

zadjii-msft replied:
Hmm, so I tried something in ffc78620e.

Instead of setting `_runtimeUseAcrylic` based on the opacity, when the opacity changes, just do the `opacity < 1.0` when we actually need it. 

That _seems_ to work... I'll play with it more but that's a start

Jaswir replied:
@zadjii-msft Jaswir took 🗡️

Jaswir replied:
@zadjii-msft Are you a DnD player?

zadjii-msft replied:
I am, but that was more of a reference to 
![image](https://github.com/microsoft/terminal/assets/18356694/3ca05658-5626-4a2a-a5d1-152dd2ac6ca9)

Jaswir replied:
I see, I didn't play much Zelda to be honest ..

zadjii-msft reviewed (commented):
Yea I suppose I'm mostly just curious about the runtime setting thing. That seems like it would get us a lot of this for free.

Sorry for the delays! Been a busy summer 😅

zadjii-msft commented on src/cascadia/TerminalSettingsModel/defaults.json:
```diff
@@ -487,6 +487,21 @@
+
+            
+        },   
+        { "command": "toggleAcrylic", "keys": "ctrl+space" },
```
actually I'm gonna block over this one

Jaswir replied:
@zadjii-msft @kalimal

@zadjii-msft @KalleOlaviNiemitalo
> actually I'm gonna block over this one

I'm sorry not sure what you mean with that phrase to block over something. I understand that it means you need to think deeply about this, is that correct? 

I think it should have a key bind for sure because:

1. I am a Terminal user and want to be able to do this with a keybind. 
2. @moonunit9 would likely agree, given the initial ask, mentioning keystrokes: "Ability to quickly toggle opacity on/off through keystrokes. It would switch between 100% and custom setting in properties."
3. It is an Acrylic Toggle so you want to be able to switch it quickly a keybind allows for this. 
4. Try it yourself and see what you prefer, I am pretty sure you actually want it to be a keybind too. But ctrl+space might not be the best choice, maybe something else would be better? 

Could we maybe ask other terminal users somewhere about their opinion on this? When I made games I'd just ask the players for stuff like this, people's choice.

zadjii-msft replied:
> > actually I'm gonna block over this one
> 
> I'm sorry not sure what you mean with that phrase: "to block over something".

Basically, in code-review parlance, "blocking" a PR is when a reviewer hits that "Request changes" button in response to an issue with the PR. By actually requesting changes, that prevents the PR from merging, even if two other people do sign off on it. 

----

It's totally fine for users to add a keybinding for this action, but I don't think we should bind one _for them_. They're always free to add their own keybinding in the settings. 

If you put it in `defaults.json` like this, then it'll bind <kbd>ctrl+space</kbd> to this action _for everyone_. Binding it here would prevent any CLI application from reading that key unless the user manually unbinds it. And <kbd>ctrl+space</kbd> is already used by different shells for a variety of things. Best we not steal it from them. 

If you change this instead to just

```json
{ "command": "toggleAcrylic" },
```

It'll still show up in the Command Palette, and in the Settings UI, so users can easily add their own keys.

Jaswir replied:
Didn't know I could add new actions for things that had no keys bound to them. 
That wasn't intuitive to me. 

I see it now: 
![image](https://github.com/microsoft/terminal/assets/15957528/b8026a79-c9e0-4787-a250-f6ac1dec808e)

Jaswir replied:
@zadjii-msft 

Updated it

Jaswir commented:
> Yea I suppose I'm mostly just curious about the runtime setting thing. That seems like it would get us a lot of this for free.
> 
> Sorry for the delays! Been a busy summer 😅

@zadjii-msft 
I am worried that when this issue gets merged you won't get notification anymore for things said in 2531 chat. And I can't use it anymore to talk in a comfortable manner about the other issue. And feel like your response doesn't adress my concern.

So I would like to know if you still get notifications when something is said inside 2531 once it's merged.

zadjii-msft commented:
> So I would like to know if you still get notifications when something is said inside https://github.com/microsoft/terminal/issues/2531 once it's merged

I get emails for every comment and on every issue, PR and discussion, and I try best to keep on top of them, no worries ☺️

Jaswir commented on src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:
```diff
@@ -571,6 +571,9 @@
     <comment>A command to change how transparent the background of the window is</comment>
   </data>
+  <data name="ToggleAcrylicCommandKey" xml:space="preserve">
+    <value>Toggle Acrylic</value>
```
Okay, how did I accept suggestions again..

Jaswir replied:
Ah I didn't know translators use those. Interesting fact

Jaswir committed:
Update src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw

Accepted Mike's suggestion comment for translators + value more consistent with codebase

Co-authored-by: Mike Griese <migrie@microsoft.com>

Jaswir committed:
removed keys from toggle acrylic command

Jaswir committed:
Fixed merge conflict appactionhandlers

Jaswir committed:
Merge branch 'microsoft:main' into jaswir/2531

Jaswir committed:
Merge branch 'microsoft:main' into jaswir/2531

Jaswir committed:
Acrylic toggle updated after Unfocused Acrylic update

Jaswir committed:
Ran code formatting

Jaswir committed:
Experimentations

Jaswir committed:
Merge branch 'microsoft:main' into jaswir/2531

Jaswir committed:
Instead of setting _runtimeUseAcrylic based on the opacity, when the opacity changes, just do the opacity < 1.0 when we actually need it.

Jaswir committed:
Main functionality added

Jaswir committed:
Merge branch 'microsoft:main' into jaswir/2531

Jaswir committed:
Fixed spelling thing

Jaswir committed:
When unfocused in Command Pallette open it should remain Acrylic!

Jaswir committed:
Fixed issue with command palette going to unfocused

Jaswir committed:
Cleaned up code for review

Jaswir committed:
Review ready

Jaswir committed:
Hacky solution

DHowett commented:
@zadjii-msft what happened with this one?
